README.md
<!-- badges: start -->
[![R-CMD-check](https://github.com/guslipkin/cipheR/actions/workflows/R-CMD-check.yaml/badge.svg)](https://github.com/guslipkin/cipheR/actions/workflows/R-CMD-check.yaml)
<!-- badges: end -->

# cipheR <img src="man/figures/logo.png" align="right" height="139" />
An R library for simple text ciphers.

## Install Process

### From CRAN

```R
install.packages("cipheR")
```

### From Source

```R
install.packages("devtools")
devtools::install_github("guslipkin/cipheR")
```


NEWS.md
## 2022-11-12 - 1.0.0
First release!


tests/testthat.R
# This file is part of the standard setup for testthat.
# It is recommended that you do not modify it.
#
# Where should you do additional test configuration?
# Learn more about the roles of various files in:
# * https://r-pkgs.org/tests.html
# * https://testthat.r-lib.org/reference/test_package.html#special-files

library(testthat)
library(cipheR)

test_check("cipheR")


tests/testthat/test-atbash.R
test_that("Letters are flipped", {
  expect_equal(atbash("az"), "za")
})

test_that("Spaces are preserved", {
  expect_equal(atbash("a b c"), "z y x")
})
test_that("Punctuation is preserved", {
  expect_equal(atbash("a.b"), "z.y")
})

test_that("Vector length is preserved", {
  expect_equal(atbash(c("a", "b", "c")), c("z", "y", "x"))
})

test_that("Must be a character vector of length greater than zero", {
  expect_error(atbash(c(TRUE)))
  expect_error(atbash(c(123)))
  expect_error(atbash(c()))
})


tests/testthat/test-caesar.R
test_that("Shift 0 warns and works", {
  expect_message(caesar("abcde", 0))
  expect_equal(caesar("abcde", 0), "abcde")
})
test_that("Shift 1 works", {
  expect_equal(caesar("abcde", 1), "bcdea")
})
test_that("Shift -1 works", {
  expect_equal(caesar("abcde", -1), "eabcd")
})

test_that("Spaces are preserved", {
  expect_equal(caesar("a b c", 1), "b c a")
})
test_that("Spaces are not preserved", {
  expect_equal(caesar("a b c", 1, preserve_spaces = FALSE), "baca ")
})

test_that("Vector length is preserved", {
  expect_equal(caesar(c("a", "b", "c"), 1), c("b", "c", "a"))
})

test_that("When no dictionary is provided, x is used", {
  expect_equal(caesar("abcde", 1), "bcdea")
})

test_that("Cannot provide both a dict and a preset", {
  expect_warning(caesar("abcde", 1, dict = "abcde", preset = "letters"))
})

test_that("The alphanumeric preset works", {
  expect_equal(caesar("abc123", 1, preset = "alphanumeric"), "bcd234")
})
test_that("The keyboard preset works", {
  expect_equal(caesar("abc123!@#", 1, preset = "keyboard"), "bcd234\"A$")
})
test_that("The letters preset works", {
  expect_equal(caesar("abcABC", 1, preset = "letters"), "bcdBCD")
})
test_that("The lowercase preset works", {
  expect_equal(caesar("abcde", 1, preset = "lowercase"), "bcdef")
})
test_that("The uppercase preset works", {
  expect_equal(caesar("ABCDE", 1, preset = "uppercase"), "BCDEF")
})
test_that("A typo in the preset gives an error", {
  expect_error(caesar("abc", 1, preset = "abc"))
})

test_that("dict must contain all values in x", {
  expect_error(caesar("abc", 1, dict = "def"))
})

test_that("x cannot be empty", {
  expect_error(caesar(c(), 1))
})
test_that("x must be a vector", {
  expect_error(caesar(list("abc"), 1))
})
test_that("x must be a character or numeric vector", {
  expect_error(caesar(c(TRUE), 1))
})



R/atbash.R
#' Encrypt or decrypt an Atbash Cipher
#'
#' @export
#'
#' @description This can be used to create (encrypt) and solve (decrypt) an
#'   Atbash Cipher. An Atbash Cipher swaps letters' places in the alphabet.
#'   Thus, 'a' becomes 'z', 'b' becomes 'y', and so on. The function does not
#'   differentiate between the two.
#'
#'   The Atbash Cipher Wikipedia entry provides more information on the methods
#'   used: [https://en.wikipedia.org/wiki/Atbash](https://en.wikipedia.org/wiki/Atbash)
#'
#' @param x A vector to be encoded or decoded.
#'
#' @returns A character vector of length one that has been transformed
atbash <- function(x) {

  if (length(x) == 0) {
    stop("Please provide a vector of length greater than zero.")
  } else if (!is.atomic(x) & !is.character(x)) {
    stop("x must be a character vector.")
  }

  x <- strsplit(x, "")

  x <- lapply(x, function(y) {
    y <-
      sapply(y, function(z) {
        if (z %in% c(letters, LETTERS)) {
          if (grepl("[A-Z]", z)) {
            z <- LETTERS[27 - which(LETTERS == z)]
          } else {
            z <- letters[27 - which(letters == z)]
          }
        }
        return(z)
      })
    y <- paste0(y, collapse = "")
    return(y)
  })

  x <- unlist(x, recursive = FALSE)

  return(x)
}


R/cipher.R
#' Encrypt or decrypt a Caesar Cipher
#'
#' @description This can be used to create (encrypt) and solve (decrypt) a
#'   Caesar cipher. The function does not differentiate between the two.
#'
#' @param x A vector to be shifted
#' @param n (Default: `1`) The number of places to shift by. This can be either
#'   positive or negative. Zero returns x as it was given to the function.
#' @param preserve_spaces (Default: `TRUE`) A boolean describing if spaces should
#'   be preserved. This is helpful when working with sentences.
#' @param dict The dictionary used for shifting. This defaults to NULL in which
#'   case a dictionary is built from the sorted unique values of x.
#' @param preset A pre-made dictionary using ASCII codes from
#'   \url{https://www.ascii-code.com/}. Note that `delete` is excluded as a
#'   character.
#'   * `NULL` (the default)
#'   * `"alphanumeric"`: ASCII characters 48:57, 65:90, and 97:122. Numbers 0-9
#'   and both uppercase and lowercase letters from the English alphabet.
#'   * `"keyboard"`: ASCII characters 32:126. The characters you'll find on a
#'   US English keyboard.
#'   * `"letters"`: ASCII characters 65:90 and 97:122. Both uppercase and
#'   lowercase letters from the English alphabet.
#'   * `"lowercase"`: ASCII characters 97:122. Lowercase letters from the
#'   English alphabet.
#'   * `"uppercase"`: ASCII characters 65:90. Uppercase letters from the
#'   English alphabet.
#'
#' @returns A character vector of length one that has been shifted.
caesar <- function(x, n = 1, preserve_spaces = TRUE, dict = NULL, preset = NULL) {

  if (length(x) == 0) {
    stop("Please provide a vector of length greater than zero to shift.")
  }

  if (!is.null(dict) & !is.null(preset)) {
    warning("Both a dict and a preset was provided, only the dict will be used.")
  }

  if (n == 0) {
    warning("Shifting by zero doesn't do anything...")
    return(x)
  }

  if (!is.character(x)) {
    x <- as.character(x)
  }

  if (length(x) == 1) {
    x <- unlist(strsplit(x, ""))
  }

  if (is.null(dict) & is.null(preset)) {
    dict <- sort(unique(x))
  } else if (!is.null(preset)) {
    if (preset == "alphanumeric") {
      dict <- rawToChar(as.raw(c(48:57, 65:90, 97:122)))
    } else if (preset == "keyboard") {
      dict <- rawToChar(as.raw(32:126))
    } else if (preset == "letters") {
      dict <- rawToChar(as.raw(c(65:90, 97:122)))
    } else if (preset == "lowercase") {
      dict <- rawToChar(as.raw(c(97:122)))
    } else if (preset == "uppercase") {
      dict <- rawToChar(as.raw(c(65:90)))
    } else {
      stop("It looks like you may have a typo in your presets. You can double-check the presets by running ?caesar in your console.")
    }
  }

  if (length(dict) == 1) {
    dict <- unlist(strsplit(dict, ""))
  }

  if (!all(x %in% dict)) {
    stop("Not all values of x are in the character set. Please choose a different character set.")
  }

  if (preserve_spaces) {
    isSpace <- which(dict == " ")
    if (length(isSpace > 0)) {
      dict <- dict[-which(dict == " ")]
    }
  }

  x <-
    sapply(x, function(y) {
      if (preserve_spaces & y == " ") {
        return(" ")
      }
      hop <- which(y == dict) + n
      hop <- hop %% length(dict)
      if (hop == 0) {
        hop <- length(dict)
      }
      return(dict[hop])
    })
  x <- paste0(x, collapse = "")

  return(x)
}


R/vigenere.R
#' Encrypt or decrypt a Vigenere Cipher
#'
#' @export
#'
#' @description This can be used to create (encrypt) and solve (decrypt) a
#'   Vigenere Cipher. A Vigenere cipher uses a table of alphabetic caesar shifts
#'   for one to twenty-six. Each letter and corresponding key value determine
#'   the grid location to choose the obfuscated letter from.
#'
#'   The Vigenere Cipher Wikipedia entry provides more information on the methods
#'   used: [https://en.wikipedia.org/wiki/Vigen%C3%A8re_cipher](https://en.wikipedia.org/wiki/Vigen%C3%A8re_cipher)
#'
#' @param x A vector to be encoded or decoded.
#' @param key A character vector of length one to use as a key
#' @param decrypt (Default: `FALSE`) The default `FALSE` will encrypt while
#'   using `TRUE` will decrypt a given value of `x`.
#' @param keep_punctuation (Default: `FALSE`) The default `FALSE` will ignore
#'   case and punctuation and return a lowercase result. `TRUE` will match the
#'   input's case and punctuation.
#'
#' @returns A character vector of length equal to x that has been transformed
vigenere <- function(x, key, decrypt = FALSE, keep_punctuation = FALSE) {

  if (length(x) == 0) {
    stop("Please provide a vector of length greater than zero for x")
  } else if (!is.atomic(x) & !is.character(x) & !is.matrix(x)) {
    stop("x must be a character vector.")
  }

  if (length(key) != 1) {
    stop("Please provide a vector of length one for key")
  } else if (!is.character(key) & (is.matrix(key) | !is.atomic(key))) {
    stop("key must be a character vector.")
  }

  if (!is.logical(decrypt)) {
    stop("decrypt must be logical")
  }
  if (!is.logical(keep_punctuation)) {
    stop("decrypt must be logical")
  }

  square <-
    suppressWarnings(matrix(sapply(0:25, function(x) {
      caesar(letters, x)
    }), 26, 26))

  key <- tolower(key)

  if (!decrypt) {
    x <- .vigenere_encrypt(x, key, square, keep_punctuation)
  } else {
    x <- .vigenere_decrypt(x, key, square, keep_punctuation)
  }

  return(x)
}

#' @keywords internal
.vigenere_encrypt <- function(x, key, square, keep_punctuation) {
  x <- lapply(x, function(x) {
    x <- unlist(strsplit(x, ""))

    if (!keep_punctuation) { x <- x[grepl("[A-z]", x)] }
    lowerX <- tolower(x)

    key <- unlist(strsplit(key, ""))
    key <- .rep_key(x, key)

    r <- sapply(key, .get_letter)
    c <- sapply(lowerX, .get_letter)

    x <- sapply(1:length(x), function(i) {
      if (c[i] == 0) {
        y <- x[i]
        print(y)
      } else {
        y <- square[r[i], c[i]]
      }

      if (keep_punctuation & grepl("[A-Z]", x[i])) { return(toupper(y)) }
      else { return(y) }
    })

    x <- paste0(x, collapse = "")
    return(x)
  })
  x <- unlist(x, recursive = FALSE)
  return(x)
}

#' @keywords internal
.vigenere_decrypt <- function(x, key, square, keep_punctuation) {
  x <- lapply(x, function(x) {
    x <- unlist(strsplit(x, ""))

    if (!keep_punctuation) { x <- x[grepl("[A-z]", x)] }
    lowerX <- tolower(x)

    key <- unlist(strsplit(key, ""))
    key <- .rep_key(x, key)

    r <- sapply(key, .get_letter)

    x <- mapply(function(r, x, lowerX) {
      i <- square[1, which(square[r, ] == lowerX)]

      if (keep_punctuation & grepl("[A-Z]", x)) { i <- toupper(i) }
      else if (keep_punctuation & length(i) == 0) { i <- x }
      return(i)
    }, r, x, lowerX)

    x <- paste0(x, collapse = "")

    return(x)
  })
  x <- unlist(x, recursive = FALSE)
  return(x)
}

#' @keywords internal
.get_letter <- function(y) {
  y <- which(letters == y)
  if (length(y) == 0) { return(0) }
  else { return(y) }
}

#' @keywords internal
.rep_key <- function(x, key) {
  rep_key <- c()
  k <- 1
  return_key <- sapply(1:length(x), function(i) {
    ki <- k %% length(key)
    if (ki == 0) { ki <- length(key) }

    if (!grepl("[A-z]", x[i])) {
      rep_key <- c(rep_key, "")
    } else {
      rep_key <- c(rep_key, key[ki])
      k <<- k + 1
    }
    return(rep_key)
  })

  return(return_key)
}


R/railfence.R
#' Encrypt or decrypt a Railfence Cipher
#'
#' @export
#'
#' @description This can be used to create (encrypt) and solve (decrypt) a
#'   Railfence Cipher. A Railfence Cipher maps each letter to a cosine wave of
#'   the specified height where each letter resides at an integer rail height.
#'
#'   The Railfence Cipher Wikipedia entry provides more information on the methods
#'   used: [https://en.wikipedia.org/wiki/Rail_fence_cipher](https://en.wikipedia.org/wiki/Rail_fence_cipher)
#'
#' @param x A vector to be encoded or decoded.
#' @param n (Default: `1`) The width of the rail to be used. A width of one will
#'   have no effect.
#' @param decrypt (Default: `FALSE`) The default `FALSE` will encrypt while
#'   using `TRUE` will decrypt a given value of `x`.
#'
#' @returns A character vector of length one that has been transformed
railfence <- function(x, n = 1, decrypt = FALSE) {

  if (length(x) == 0) {
    stop("Please provide a vector of length greater than zero")
  } else if (!is.atomic(x) & !is.character(x)) {
    stop("x must be a character vector.")
  }

  if (length(n) != 1 || n < 1 || n %% 1 != 0) {
    stop("n must be a single integer greater than or equal to 1")
  }

  if (!is.logical(decrypt)) {
    stop("decrypt must be logical")
  }

  if (!decrypt) {
    x <- .railfence_encrypt(x, n)
  } else {
    x <- .railfence_decrypt(x, n)
  }

  return(x)
}

#' @keywords internal
.railfence_encrypt <- function(x, n) {
  x <- strsplit(x, "")

  x <- lapply(x, function(y) {
    nx <- length(y)
    z <- c(1, rep_len(c(2:n, (n - 1):1), nx - 1))
    y <- data.frame("x" = y, "pos" = z)
    y <- y[order(y$pos), ]
    y <- paste0(y$x, collapse = "")
    return(y)
  })

  x <- unlist(x, recursive = FALSE)

  return(x)
}

#' @keywords internal
.railfence_decrypt <- function(x, n) {
  x <-
    lapply(x, function(s) {
      s <- unlist(strsplit(s, ""))
      m <- matrix(NA, n, length(s))
      j <- 1
      dir <- 1
      for (i in 1:length(s)) {
        m[j, i] <- "*"

        if (dir == 1) {
          if (j == n) {
            dir <- -1
            j <- j - 1
          } else {
            j <- j + 1
          }
        } else {
          if (j == 1) {
            dir <- 1
            j <- j + 1
          } else {
            j <- j - 1
          }
        }
      }
      k <- 1
      for (j in 1:nrow(m)) {
        for (i in 1:ncol(m)) {
          if (!is.na(m[j, i])) {
            m[j, i] <- s[k]
            k <- k + 1
          }
        }
      }
      m <- as.vector(m)
      m <- paste(m[!is.na(m)], collapse = "")

      return(m)
    })

  x <- unlist(x, recursive = FALSE)

  return(x)
}


R/running_key.R
#' Encrypt or decrypt a Running Key Vigenere Cipher
#'
#' @export
#'
#' @description This can be used to create (encrypt) and solve (decrypt) a
#'   Running Key Vigenere Cipher. A Vigenere cipher uses a table of alphabetic
#'   caesar shifts for one to twenty-six. The key is made to have an equal
#'   length to the text by adding the first letters of the text to the key. Each
#'   letter and corresponding key value determine the grid location to choose
#'   the obfuscated letter from.
#'
#'   The Running Key Cipher Wikipedia entry provides more information on the
#'   methods used: [https://en.wikipedia.org/wiki/Running_key_cipher](https://en.wikipedia.org/wiki/Running_key_cipher)
#'
#' @param x A vector to be encoded or decoded.
#' @param key A character vector of length one to use as a key
#' @param decrypt (Default: `FALSE`) The default `FALSE` will encrypt while
#'   using `TRUE` will decrypt a given value of `x`.
#' @param keep_punctuation (Default: `FALSE`) The default `FALSE` will ignore
#'   case and punctuation and return a lowercase result. `TRUE` will match the
#'   input's case and punctuation.
#'
#' @returns A character vector of length equal to x that has been transformed
running_key <- function(x, key, decrypt = FALSE, keep_punctuation = FALSE) {
  if (length(x) != 1) {
    stop("Please provide a vector of length one for x")
  } else if (!is.atomic(x) & !is.character(x) & !is.matrix(x)) {
    stop("x must be a character vector.")
  }

  if (length(key) != 1) {
    stop("Please provide a vector of length one for key")
  } else if (!is.character(key) & (is.matrix(key) | !is.atomic(key))) {
    stop("key must be a character vector.")
  }

  y <- unlist(strsplit(x, ""))
  y <- y[grepl("[A-z]", y)]

  k <- unlist(strsplit(key, ""))
  k <- k[grepl("[A-z]", k)]

  if(length(k) >= length(y)) {
    x <- vigenere(x, paste0(k, collapse = ""), decrypt = decrypt, keep_punctuation = keep_punctuation)
    return(x)
  } else {
    stop(paste("The key must have an equal or greater number of letters to the text. The key has",
               length(k), "characters and the message has", length(y), "characters."))
  }
}


R/caesar.R
#' Encrypt or decrypt a Caesar Cipher
#'
#' @export
#'
#' @description This can be used to create (encrypt) and solve (decrypt) a
#'   Caesar cipher. The function does not differentiate between the two.
#'
#'   The Caesar Cipher Wikipedia entry provides more information on the methods
#'   used: [https://en.wikipedia.org/wiki/Caesar_cipher](https://en.wikipedia.org/wiki/Caesar_cipher)
#'
#' @param x A vector to be shifted
#' @param n (Default: `1`) The number of places to shift by. This can be either
#'   positive or negative. Zero returns x as it was given to the function.
#' @param preserve_spaces (Default: `TRUE`) A boolean describing if spaces should
#'   be preserved. This is helpful when working with sentences.
#' @param dict The dictionary used for shifting. This defaults to NULL in which
#'   case a dictionary is built from the sorted unique values of x.
#' @param preset A pre-made dictionary using ASCII codes from
#'   \url{https://www.ascii-code.com/}. Note that `delete` is excluded as a
#'   character.
#'   * `NULL` (the default)
#'   * `"alphanumeric"`: ASCII characters 48:57, 65:90, and 97:122. Numbers 0-9
#'   and both uppercase and lowercase letters from the English alphabet.
#'   * `"keyboard"`: ASCII characters 32:126. The characters you'll find on a
#'   US English keyboard.
#'   * `"letters"`: ASCII characters 65:90 and 97:122. Both uppercase and
#'   lowercase letters from the English alphabet.
#'   * `"lowercase"`: ASCII characters 97:122. Lowercase letters from the
#'   English alphabet.
#'   * `"uppercase"`: ASCII characters 65:90. Uppercase letters from the
#'   English alphabet.
#'
#' @returns A character vector of length one that has been shifted.
caesar <- function(x, n = 1, preserve_spaces = TRUE, dict = NULL, preset = NULL) {

  # Catching errors
  if (length(x) == 0) {
    stop("Please provide a vector of length greater than zero to shift.")
  } else if (!is.atomic(x) | !(is.character(x) | is.numeric(x))) {
    stop("x must be a numeric or character vector.")
  }

  if (!is.null(dict) & !is.null(preset)) {
    warning("Both a dict and a preset was provided, only the dict will be used.")
  }

  if (n == 0) {
    message("Shifting by zero doesn't do anything...")
    return(x)
  }

  if (!is.character(x)) {
    x <- as.character(x)
  }

  x <- strsplit(x, "")
  unlistX <- unlist(x)

  # Set the dictionary as needed
  if (is.null(dict) & is.null(preset)) {
    dict <- sort(unique(unlistX))
  } else if (!is.null(preset)) {
    if (preset == "alphanumeric") {
      dict <- rawToChar(as.raw(c(48:57, 65:90, 97:122)))
    } else if (preset == "keyboard") {
      dict <- rawToChar(as.raw(32:126))
    } else if (preset == "letters") {
      dict <- rawToChar(as.raw(c(65:90, 97:122)))
    } else if (preset == "lowercase") {
      dict <- rawToChar(as.raw(c(97:122)))
    } else if (preset == "uppercase") {
      dict <- rawToChar(as.raw(c(65:90)))
    } else {
      stop("It looks like you may have a typo in your presets. You can double-check the presets by running ?caesar in your console.")
    }
  }

  if (length(dict) == 1) {
    dict <- unlist(strsplit(dict, ""))
  }

  if (!all(unlistX %in% dict)) {
    stop("Not all values of x are in the character set. Please choose a different character set.")
  }

  x <-
    lapply(x, function(y) {
      # We need to preserve spaces as requested
      if (preserve_spaces) {
        isSpace <- which(dict == " ")
        if (length(isSpace > 0)) {
          dict <- dict[-which(dict == " ")]
        }
      }

      # Do the shifting
      y <-
        sapply(y, function(z) {
          if (preserve_spaces & z == " ") {
            return(" ")
          }
          hop <- which(z == dict) + n
          hop <- hop %% length(dict)
          if (hop == 0) {
            hop <- length(dict)
          }
          return(dict[hop])
        })
      y <- paste0(y, collapse = "")
      return(y)
    })

  x <- unlist(x, recursive = FALSE)

  return(x)
}
